src/Microsoft.Diagnostics.EventFlow.Core/Implementations/LossReportingPropagatorBlock.cs

```diff
@@ -0,0 +1,209 @@
+        }
+
+        public Task Completion
+        {
+            get
+            {
+                if (this.propagateCompletion)
+                {
+                    Debug.Assert(this.target != null);
+                    return this.target.Completion;
+                }
+                else
+                {
+                    return this.completionSource.Value.Task;
+                }
+            }
+        }
+
+        public void Complete()
+        {
+            this.completed = true;
+
+            if (this.propagateCompletion)
+            {
+                Debug.Assert(this.target != null);
+                this.target.Complete();
+            }
+            else
+            {
+                this.completionSource.Value.TrySetResult(true);
+            }
+        }
+
+        public void Fault(Exception exception)
+        {
+            this.completed = true;
+
+            if (this.propagateCompletion)
+            {
+                Debug.Assert(this.target != null);
+                this.target.Fault(exception);
+            }
+            else
+            {
+                this.completionSource.Value.TrySetException(exception);
+            }
+        }
+
+        public IDisposable LinkTo(ITargetBlock<T> target, DataflowLinkOptions linkOptions)
+        {
+            Requires.NotNull(target, nameof(target));
+            if (this.target != null)
+            {
+                throw new InvalidOperationException($"{nameof(LossReportingPropagatorBlock<T>)} only supports a single target");
+            }
+
+            this.target = target;
+            if (linkOptions != null)
+            {
+                // Note DataflowLinkOptions.Append does not matter given that we are working with a single target only 
+                // and we will not enforce DataflowLinkOptions.MaxMessages
+                this.propagateCompletion = linkOptions.PropagateCompletion;
+            }
+
+            // We do not support re-linking
+            return new EmptyDisposable();
+        }
+
+        public DataflowMessageStatus OfferMessage(DataflowMessageHeader messageHeader, T messageValue, ISourceBlock<T> source, bool consumeToAccept)
+        {
+            VerifySourceInitiatedOperation(source);
+
+            if (this.completed)
+            {
+                return DataflowMessageStatus.DecliningPermanently;
+            }
+
+            var targetResponse = this.target.OfferMessage(messageHeader, messageValue, this, consumeToAccept);
+            switch(targetResponse)
+            {
+                case DataflowMessageStatus.Accepted:
+                    // Great, nothing to do
+                    break;
+
+                case DataflowMessageStatus.Declined:
+                case DataflowMessageStatus.DecliningPermanently:
+                case DataflowMessageStatus.NotAvailable:
+                    this.healthReporter.ReportThrottling();
```

Comment:
My aim with this implementation was to create something that makes sense for variety of situations, not just in the context of being a downstream block for BroadcastBlock. It also makes it easier to verify if the block understands the full DataFlow protocol with all the possible responses handled.